engine/runtime/src/main/java/org/enso/interpreter/runtime/callable/atom/AtomConstructor.java

```diff
@@ -32,7 +33,7 @@
 /** A representation of an Atom constructor. */
 @ExportLibrary(InteropLibrary.class)
 @ExportLibrary(MethodDispatchLibrary.class)
-public final class AtomConstructor implements TruffleObject {
+public class AtomConstructor implements TruffleObject {
```

Comment:
Can't do that. For any runtime value (that is ever sometimes pattern matched on in specializations), we need to be a leaf of the type hierarchy, otherwise `instanceof` checks become much slower  – a leaf check is just `==` on class ptrs, with this change you need to actually traverse the class hierarchy. So the `BuiltinAtomConstructor` needs to become something separate, or you'll have to pull the functionality into this class, with boolean flags and other such niceties.